### PR TITLE
Move GItHub App creation routes to HTTP API

### LINKED ISSUE #59771: GitHub App creation fails using http-header auth
- **Sourcegraph version:** 5.2.4 

#### Steps to reproduce:

1. Setup the http-header auth provider.
2. Create a GitHub app to trigger the error.

#### Expected behavior:
The app creation flow continues.

#### Actual behavior:
There is an error. User must be site admin.

Screenshot 2024-01-18 at 10 44 55 PM

If you would like immediate help on this, please email support@sourcegraph.com (you can still create the issue, but there are no [SLAs](https://sourcegraph.com/support/) on issues like there are for support requests).

### PR DESCRIPTION
**Pull request by pjlast**
Closes #59771 

This PR moves the HTTP endpoints related to GitHub App creation to the rest of the HTTP API.

These routes lived in the `/.auth/...` endpoints, which is completely wrong since these endpoints aren't related to any sort of auth. This also lead to a weird situation where `http-header` auth failed when trying to create a GitHub App, because the GitHub App endpoints require authentication and the user would not be authenticated yet at this point.

## Test plan

Manual tests and unit/integration tests still passing.

### CODE REVIEW

File: cmd/frontend/internal/app/app.go [hunk 1]
```diff
@@ -78,10 +78,6 @@ func NewHandler(db database.DB, logger log.Logger, githubAppSetupHandler http.Ha
 	// Ping retrieval
 	r.Get(router.LatestPing).Handler(trace.Route(latestPingHandler(db)))
 
-	// Sourcegraph GitHub App setup (Cloud and on-prem)
-	r.Get(router.SetupGitHubAppCloud).Handler(trace.Route(githubAppSetupHandler))
-	r.Get(router.SetupGitHubApp).Handler(trace.Route(githubAppSetupHandler))
```
Comment (pjlast) [inline cmd/frontend/internal/app/app.go:83]: These were unused and were from the very old GitHub App integration

File: cmd/frontend/internal/httpapi/httpapi.go [hunk 2]
```diff
@@ -136,39 +138,43 @@ func NewHandler(
...
-	m.Path("/bitbucket-server-webhooks").Methods("POST").Handler(trace.Route(webhookMiddleware.Logger(handlers.BatchesBitbucketServerWebhook)))
-	m.Path("/bitbucket-cloud-webhooks").Methods("POST").Handler(trace.Route(webhookMiddleware.Logger(handlers.BatchesBitbucketCloudWebhook)))
+	m.Path("/github-webhooks").Methods("POST").Handler(webhookMiddleware.Logger(&gitHubWebhook))
+	m.Path("/gitlab-webhooks").Methods("POST").Handler(webhookMiddleware.Logger(handlers.BatchesGitLabWebhook))
+	m.Path("/bitbucket-server-webhooks").Methods("POST").Handler(webhookMiddleware.Logger(handlers.BatchesBitbucketServerWebhook))
+	m.Path("/bitbucket-cloud-webhooks").Methods("POST").Handler(webhookMiddleware.Logger(handlers.BatchesBitbucketCloudWebhook))
+
+	// GitHub App handler
+	gitHubAppRouter := m.PathPrefix("/githubapp/").Subrouter()
+	githubapp.SetupGitHubAppRoutes(gitHubAppRouter, db)
```
Comment (pjlast) [inline cmd/frontend/internal/httpapi/httpapi.go:152]: This registers all of the GitHub App-specific routes on the Subrouter. I find this a million times more pleasant than the `handlers.X` setup hell.

If this ends up being an issue for testing or something I'll revisit this.

File: cmd/frontend/internal/httpapi/httpapi.go [hunk 3]
```diff
@@ -99,6 +100,7 @@ func NewHandler(
 
 	m := mux.NewRouter().PathPrefix("/.api/").Subrouter()
 	m.StrictSlash(true)
+	m.Use(trace.Route)
```
[the remark this answers is not in the document]
Comment (eseliger) [reply cmd/frontend/internal/httpapi/httpapi.go:103]: nice!

File: cmd/frontend/internal/app/app.go (hunk 1 above)
Comment (eseliger) [reply cmd/frontend/internal/app/app.go:83]: nice find

File: client/web/src/components/gitHubApps/CreateGitHubAppPage.tsx [hunk 4]
```diff
@@ -97,8 +97,8 @@ export const CreateGitHubAppPage: FC<CreateGitHubAppPageProps> = ({
                 name: name.trim(),
                 url: originURL,
                 hook_attributes: webhookURL ? { url: webhookURL } : undefined,
-                redirect_url: new URL('/.auth/githubapp/redirect', originURL).href,
-                setup_url: new URL('/.auth/githubapp/setup', originURL).href,
+                redirect_url: new URL('/.api/githubapp/redirect', originURL).href,
+                setup_url: new URL('/.api/githubapp/setup', originURL).href,
```
Comment (eseliger) [inline client/web/src/components/gitHubApps/CreateGitHubAppPage.tsx:101]: I don't really know how this works, but can this change break existing app installations?

**Review by eseliger — APPROVED** [review]

File: client/web/src/components/gitHubApps/CreateGitHubAppPage.tsx (hunk 4 above)
Comment (pjlast) [reply client/web/src/components/gitHubApps/CreateGitHubAppPage.tsx]: Nope, this is only relevant while the App is being created. Once created, this doesn't affect anything.

**Pull request merged**